ckks/encoder.go

```diff
@@ -18,49 +18,161 @@ const GaloisGen uint64 = 5
-
-	EncodeDiagMatrixBSGSAtLvl(level int, vector map[int][]complex128, scale, maxM1N2Ratio float64, logSlots int) (matrix PtDiagMatrix)
-	EncodeDiagMatrixAtLvl(level int, vector map[int][]complex128, scale float64, logSlots int) (matrix PtDiagMatrix)
-	EncodeDiagonal(logSlots, level int, scale float64, m []complex128) (vecQP rlwe.PolyQP)
 
+	// Slots Encoding
+
+	// Encode encodes a set of values on the target plaintext.
+	// This method is identical to "EncodeSlots".
+	// Encoding is done at the level and scale of the plaintext.
+	// User must ensure that 1 <= len(values) <= 2^logSlots < 2^logN.
+	// values.(type) can be either []complex128 of []float64.
+	// The imaginary part of []complex128 will be discarded if ringType == ring.ConjugateInvariant.
+	// Returned plaintext is always in the NTT domain.
+	Encode(values interface{}, plaintext *Plaintext, logSlots int)
+
+	// EncodeNew encodes a set of values on a new plaintext.
+	// This method is identical to "EncodeSlotsNew".
+	// Encoding is done at the provided level and with the provided scale.
+	// User must ensure that 1 <= len(values) <= 2^logSlots < 2^logN.
+	// values.(type) can be either []complex128 of []float64.
+	// The imaginary part of []complex128 will be discarded if ringType == ring.ConjugateInvariant.
+	// Returned plaintext is always in the NTT domain.
+	EncodeNew(values interface{}, level int, scale float64, logSlots int) (plaintext *Plaintext)
+
+	// EncodeSlots encodes a set of values on the target plaintext.
+	// Encoding is done at the level and scale of the plaintext.
+	// User must ensure that 1 <= len(values) <= 2^logSlots < 2^logN.
+	// values.(type) can be either []complex128 of []float64.
+	// The imaginary part of []complex128 will be discarded if ringType == ring.ConjugateInvariant.
+	// Returned plaintext is always in the NTT domain.
+	EncodeSlots(values interface{}, plaintext *Plaintext, logSlots int)
+
+	// EncodeSlotsNew encodes a set of values on a new plaintext.
+	// Encoding is done at the provided level and with the provided scale.
+	// User must ensure that 1 <= len(values) <= 2^logSlots < 2^logN.
+	// values.(type) can be either []complex128 of []float64.
+	// The imaginary part of []complex128 will be discarded if ringType == ring.ConjugateInvariant.
+	// Returned plaintext is always in the NTT domain.
+	EncodeSlotsNew(values interface{}, level int, scale float64, logSlots int) (plaintext *Plaintext)
+
+	// EncodeSlotsQP encodes a set of values on the target rlwe.polyQP.
+	// This method is identical to "EncodeSlots".
+	// Encoding is done at the level of the rlwe.polyQP, and provided scale.
+	// User must ensure that 1 <= len(values) <= 2^logSlots < 2^logN.
+	// values.(type) can be either []complex128 of []float64.
+	// The imaginary part of []complex128 will be discarded if ringType == ring.ConjugateInvariant.
+	// Returned rlwe.polyQP is always in the NTT domain.
+	EncodeSlotsQP(values interface{}, vecQP rlwe.PolyQP, scale float64, logSlots int)
```

Comment:
Encoding plaintexts in QP seems very specific. Should it be part of a more specific interface instead of making `Encoder` more complex ?